[Feature] Additional features for error mitigation protocols (#56)
pasqal-io/qadence-protocols: docs/error_mitigation/readout_mitigation.md

from scipy.sparse.linalg import gmres

n_qubits = 10

# Prepare a probability distribution with sparsity
exact_prob = np.random.rand(2 ** n_qubits)
exact_prob[2 ** (n_qubits // 2):] = 0
exact_prob = 0.90 * exact_prob + 0.1 * np.ones(2 ** n_qubits) / 2 ** n_qubits
exact_prob /= sum(exact_prob)
np.random.shuffle(exact_prob)

# Create an observed probability distribution with thresholded values
observed_prob = np.array(exact_prob, copy=True)
observed_prob[exact_prob < 1 / 2 ** n_qubits] = 0
observed_prob /= sum(observed_prob)

# Convert the observed probability distribution into a sparse matrix
input_csr_matrix = csr_matrix(observed_prob, shape=(1, 2**n_qubits)).T

# Print the binary representation of states with nonzero probabilities
print({
    bin(x)[2:].zfill(n_qubits): np.round(input_csr_matrix[x, 0], 3)
    for x in input_csr_matrix.nonzero()[0]
}) # markdown-exec: hide

# Compute and display the percentage of nonzero entries
filling_percentage = len(input_csr_matrix.nonzero()[0]) / 2**n_qubits
print(f"Filling percentage: {filling_percentage:.6f} %") # markdown-exec: hide
```

> We have constructed a probability distribution over a small subspace of bitstrings, leveraging a `csr_matrix` for efficient storage and computation. Now, we apply `MTHREE` to mitigate errors in the probability distribution. Within `MTHREE`, sparsity can be further improved by incorporating the Hamming distance approach. This method considers only the noise matrix elements corresponding to quantum states within a specified Hamming distance of the correct state. This feature can be activated by setting the `hamming_dist` option.

Reviewer: ```suggestion
We have constructed a probability distribution over a small subspace of bitstrings, leveraging a `csr_matrix` for efficient storage and computation. Now, we apply `MTHREE` to mitigate errors in the probability distribution. Within `MTHREE`, sparsity can be further improved by incorporating the Hamming distance approach. This method considers only the noise matrix elements corresponding to quantum states within a specified Hamming distance of the correct state. This feature can be used by setting the `hamming_dist` option.
```

Nice :)

    mle_solve,
    matrix_inv,
    tensor_rank_mult
)

# Generate noise transition matrices for each qubit
noise_matrices = []
for qubit_idx in range(n_qubits):
    transition_prob_a, transition_prob_b = np.random.rand(2) / 8
    transition_matrix = np.array([[1 - transition_prob_a, transition_prob_a],
                                  [transition_prob_b, 1 - transition_prob_b]]).T  # Ensure column sums to 1
    noise_matrices.append(transition_matrix)

> # Compute the subspace confusion matrix using noise transition matrices. We set hamming distance for filtering out noisy states that are far from the correct state

Reviewer: ```suggestion
# Compute the subspace confusion matrix using noise transition matrices. We set the hamming distance for filtering out noisy states that are far from the correct state
```

subspace_confusion_matrix = normalized_subspace_kron(noise_matrices, observed_prob.nonzero()[0], hamming_dist=1)

> # Apply GMRES (Generalized Minimal Residual Method) to correct the probability distribution using MTHREE. Then we apply Maximum Likelihood Estimation (MLE) to ensure valid probability distribution.

Reviewer: ```suggestion
# Apply GMRES (Generalized Minimal Residual Method) to correct the probability distribution using MTHREE. Then we apply Maximum Likelihood Estimation (MLE) to ensure the validity of the probability distribution.
```
Do you mean that you check for normality ?

Author: Yes correct. GMRES does not ensure normality and we need mle to do it.

corrected_prob_mthree_mle = mle_solve(gmres(subspace_confusion_matrix, input_csr_matrix.toarray())[0])

> # Use tensor rank multiplication to apply the inverse noise matrices to the observed probability distribution, followed by MLE correction

Reviewer: ```suggestion
# Next, we use tensor rank multiplication to apply the inverse noise matrices to the observed probability distribution, followed by the same MLE correction
```

inverse_noise_matrices = list(map(matrix_inv, noise_matrices))
corrected_prob_inverse_mle = mle_solve(tensor_rank_mult(inverse_noise_matrices, observed_prob))

> # Compute the Wasserstein distance between the two corrected probability distributions

Reviewer: ```suggestion
# Finally, we  compute the Wasserstein distance between the two corrected probability distributions
```

wasserstein_dist = wasserstein_distance(corrected_prob_mthree_mle, corrected_prob_inverse_mle)
print(f"Wasserstein distance between the two distributions: {wasserstein_dist}")  # markdown-exec: hide
```

> We have used `wasserstein_distance` instead of `kl_divergence` as many bistrings have 0 probabilities. If the expected solution lies outside the space of observed bitstrings, `MTHREE` will fail. We next look at majority voting to circument this problem when the expected output is a single bitstring.

Reviewer: ```suggestion
We have used `wasserstein_distance` instead of `kl_divergence` as many bistrings have 0 probabilities. 
```

Sure but it doesn't tell why the kl divergence would fail or be less accurate when bitstrings have 0 probabilities.

Author: Thanks. Updated the explanation in docs. It's basically because KL includes putting 0 value at the denominator and may diverge.

    mle_solve,
    matrix_inv,
    tensor_rank_mult
)

# Generate noise transition matrices for each qubit
noise_matrices = []
for qubit_idx in range(n_qubits):
    transition_prob_a, transition_prob_b = np.random.rand(2) / 8
    transition_matrix = np.array([[1 - transition_prob_a, transition_prob_a],
                                  [transition_prob_b, 1 - transition_prob_b]]).T  # Ensure column sums to 1
    noise_matrices.append(transition_matrix)

> # Hamming distance for filtering out noisy states that are far from the correct state

Reviewer: Add this comment below.

> hamming_dist = 1

Reviewer: Remove.

# Compute the subspace confusion matrix using noise transition matrices

> subspace_confusion_matrix = normalized_subspace_kron(noise_matrices, observed_prob.nonzero()[0], hamming_dist)

Reviewer: ```suggestion
subspace_confusion_matrix = normalized_subspace_kron(noise_matrices, observed_prob.nonzero()[0], hamming_dist=1)
```

# Apply GMRES (Generalized Minimal Residual Method) to correct the probability distribution using MTHREE
corrected_prob_mthree_gmres = gmres(subspace_confusion_matrix, input_csr_matrix.toarray())[0]

> corrected_prob_mthree_mle = mle_solve(corrected_prob_mthree_gmres)  # Apply Maximum Likelihood Estimation (MLE)

Reviewer: It is a bit confusing as it feels like you're correcting once with gmres and once again with mle. Maybe the naming should change a little ?

Author: Yes it is done in two steps. GMRES(from scipy) solves the linear system, while MLE normalizes the probability distribution. GMRES alone may not ensure a valid probability distribution, so MLE adjusts negative values and rescales probabilities. I'll add this information to the documentation and improve the naming to avoid any misunderstandings.

Author: Oh I'll also follow comments in the below

# Use tensor rank multiplication to apply the inverse noise matrices to the observed probability distribution, followed by MLE correction
inverse_noise_matrices = list(map(matrix_inv, noise_matrices))

> corrected_prob_inverse_mle = mle_solve(tensor_rank_mult(inverse_noise_matrices, observed_prob))

Reviewer: I would use the exact same workflow as above to illustrate that `mle_solve` takes corrected probas and only applies MLE on top.

# Compute the Wasserstein distance between the two corrected probability distributions
wasserstein_dist = wasserstein_distance(corrected_prob_mthree_mle, corrected_prob_inverse_mle)
print(f"Wasserstein distance between the two distributions: {wasserstein_dist}")  # markdown-exec: hide
```

> We have used `wasserstein_distance` instead of `kl_divergence` as many of the bistrings have 0 probabilites. If the expected solution lies outside the space of observed bitstrings, `MTHREE` will fail. We next look at majority voting to circument this problem when the expected output is a single bitstring.

Reviewer: ```suggestion
We have used `wasserstein_distance` instead of `kl_divergence` as many bitstrings have 0 probabilities. If the expected solution lies outside the space of observed bitstrings, `MTHREE` will fail. We next look at majority voting to circument this problem when the expected output is a single bitstring.
```

distance = wasserstein_distance(p_corr_mthree_gmres_mle, p_corr_inv_mle)
print(f"Wasserstein distance between the 2 distributions: {distance}")  # markdown-exec: hide

```

We have used `wasserstein_distance` instead of `kl_divergence` as many of the bistrings have 0 probabilites. If the expected solution lies outside the space of observed bitstrings, `MTHREE` will fail. We next look at majority voting to circument this problem when the expected output is a single bitstring.

Additionally, we can further enhance sparsity by integrating the Hamming distance approach into the `MTHREE` method. This method utilizes only the noise matrix values that fall within the specified Hamming distance of the correct quantum state. This functionality can be enabled by setting the `ham_dist` option.

```python exec="on" source="material-block" session="m3" result="json"
from qadence_protocols.mitigations.readout import ham_dist_redistribution

confusion_matrix_subspace = normalized_subspace_kron(noise_matrices, observed_prob.nonzero()[0])

# we consider a small hamming distance for this method and set it to 3

> confusion_matrix_subspace_ham = ham_dist_redistribution(

Reviewer: Instead of another function, would it make sense to integrate it into `normalized_subspace_kron` ? If I understand correctly, you use this latter function to generate a confusion matrix. But maybe if there is an extra `ham_dist` parameter that could trigger the subspace redistribution without having to rely on some logic defined elsewhere ?

Author: Yes, I revised the code to execute it in `normalized_subspace_kron` with a parameter trigger. Thanks.

                    confusion_matrix_subspace, ham_dist=3
                )

> p_corr_mthree_gmres_ham = gmres(confusion_matrix_subspace_ham, input_csr.toarray())[0]

Reviewer: hm these variables names are a bit hard to understand. Could you please use more meaningful names or comment what you're doing here ?

### Twirl mitigation

> This protocol makes use of all possible so-called twirl operations to average out the effect of readout errors into an effective scaling. The twirl operation consists of using bit flip operators before the measurement and after the measurement is obtained[^5]. The number of twirl operations can be reduced through random sampling with `twirl_samples` option. The method is exact in that it requires no calibration which might be prone to errors of modelling.

Reviewer: ```suggestion
This protocol makes use of all possible so-called twirl operations to average out the effect of readout errors into an effective scaling. The twirl operation consists of using bit flip operators before and after the measurement [^5]. The number of twirl operations can be reduced through random sampling with the `twirl_samples` option. The method is exact in that it requires no calibration which might be prone to modelling errors.
```

mitigate = Mitigations(protocol=Mitigations.TWIRL)
expectation_mitigated = mitigate(noise=noise, model=noisy_model)

> # We set a number of qubits as the sample count. The number of twirl_samples can range from 1 to the maximum number of qubit index combinations. For example,  In general, using a higher number of samples can improve accuracy.

Reviewer: ```suggestion
# We set a number of qubits as the sample count. The number of twirl_samples can range from 1 to the maximum number of qubit index combinations. For example, using a higher number of samples can improve accuracy.
```

mitigate = Mitigations(protocol=Mitigations.TWIRL)
expectation_mitigated = mitigate(noise=noise, model=noisy_model)

> # we set a number of qubits as the sample count. It can be changed for higher accuracy.

Reviewer: Changed to what value ? Higher ? Lower ?

Author: Higher and it has upper bound in sample count. I added this in the docs.
